// @terra-money/terra.js/dist/core/oracle/params.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OracleParamChanges = void 0;
var convert_1 = require("../../util/convert");
var numeric_1 = require("../numeric");
var OracleParamChanges;
(function (OracleParamChanges) {
    OracleParamChanges.ConversionTable = {
        oracle: {
            VotePeriod: [convert_1.Convert.toNumber, convert_1.Convert.toFixed],
            VoteThreshold: [convert_1.Convert.toDec, convert_1.Convert.toString],
            RewardBand: [convert_1.Convert.toDec, convert_1.Convert.toString],
            RewardDistributionWindow: [convert_1.Convert.toNumber, convert_1.Convert.toFixed],
            Whitelist: [
                function (c) {
                    return c.map(function (v) { return ({
                        name: v.name,
                        tobin_tax: new numeric_1.Dec(v.tobin_tax),
                    }); });
                },
                function (c) {
                    return c.map(function (v) { return ({
                        name: v.name,
                        tobin_tax: v.tobin_tax.toString(),
                    }); });
                },
            ],
            SlashFraction: [convert_1.Convert.toDec, convert_1.Convert.toString],
            SlashWindow: [convert_1.Convert.toNumber, convert_1.Convert.toFixed],
            MinValidPerWindow: [convert_1.Convert.toDec, convert_1.Convert.toString],
        },
    };
})(OracleParamChanges = exports.OracleParamChanges || (exports.OracleParamChanges = {}));
