// @terra-money/terra.js/dist/core/treasury/params.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TreasuryParamChanges = void 0;
var PolicyConstraints_1 = require("./PolicyConstraints");
var convert_1 = require("../../util/convert");
var TreasuryParamChanges;
(function (TreasuryParamChanges) {
    TreasuryParamChanges.ConversionTable = {
        treasury: {
            TaxPolicy: [PolicyConstraints_1.PolicyConstraints.fromData, convert_1.Convert.toData],
            RewardPolicy: [PolicyConstraints_1.PolicyConstraints.fromData, convert_1.Convert.toData],
            SeigniorageBurdenTarget: [convert_1.Convert.toDec, convert_1.Convert.toString],
            MiningIncrement: [convert_1.Convert.toDec, convert_1.Convert.toString],
            WindowShort: [convert_1.Convert.toNumber, convert_1.Convert.toFixed],
            WindowLong: [convert_1.Convert.toNumber, convert_1.Convert.toFixed],
            WindowProbation: [convert_1.Convert.toNumber, convert_1.Convert.toFixed],
        },
    };
})(TreasuryParamChanges = exports.TreasuryParamChanges || (exports.TreasuryParamChanges = {}));
