// @terra-money/terrain/lib/commands/code/store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const terra_js_1 = require("@terra-money/terra.js");
const config_1 = require("../../config");
const deployment_1 = require("../../lib/deployment");
const signer_1 = require("../../lib/signer");
class CodeStore extends command_1.Command {
    async run() {
        const { args, flags } = this.parse(CodeStore);
        const connections = config_1.loadConnections(flags["config-path"]);
        const config = config_1.loadConfig(flags["config-path"]);
        const conf = config(flags.network, args.contract);
        // @ts-ignore
        const lcd = new terra_js_1.LCDClient(connections(flags.network));
        const signer = signer_1.getSigner({
            network: flags.network,
            signerId: flags.signer,
            keysPath: flags["keys-path"],
            lcd,
        });
        await deployment_1.storeCode({
            conf,
            noRebuild: flags["no-rebuild"],
            contract: args.contract,
            signer,
            network: flags.network,
            refsPath: flags["refs-path"],
            lcd: lcd,
            codeId: flags["code-id"],
        });
    }
}
exports.default = CodeStore;
CodeStore.description = "Store code on chain.";
CodeStore.flags = {
    "no-rebuild": command_1.flags.boolean({ default: false }),
    network: command_1.flags.string({ default: "localterra" }),
    "config-path": command_1.flags.string({ default: "./config.terrain.json" }),
    "refs-path": command_1.flags.string({ default: "./refs.terrain.json" }),
    "keys-path": command_1.flags.string({ default: "./keys.terrain.js" }),
    "code-id": command_1.flags.integer({}),
    signer: command_1.flags.string({ required: true }),
};
CodeStore.args = [{ name: "contract", required: true }];
